project/src/main/java/com/google/sps/data/VisionLibrary.java

```diff
@@ -0,0 +1,330 @@
+    List<ColorInfo> colorAnnotations = colors.getColorsList();
+
+    // Gets a list of all objects found in the image.
+    List<LocalizedObjectAnnotation> objectAnnotations = imageResponse.getLocalizedObjectAnnotationsList();
+
+    // Finds any logos that are in the image.
+    List<EntityAnnotation> logoAnnotations = imageResponse.getLogoAnnotationsList();
+
+    List<ImageLabel> genericLabels = new ArrayList<>();
+    for (EntityAnnotation label : labelAnnotations) {
+      genericLabels.add(new ImageLabel(label.getDescription(), label.getScore()));
+    }
+
+    List<ImageLabel> webLabels = new ArrayList<>();
+    for (WebEntity label : webLabelAnnotations) {
+      webLabels.add(new ImageLabel(label.getDescription(), label.getScore()));
+    }
+
+    List<String> webBestLabels = new ArrayList<>();
+    for (WebLabel label : webBestLabelAnnotations) {
+      webBestLabels.add(label.getLabel());
+    }
+
+    List<String> webUrls = new ArrayList<>();
+    for (WebPage page : webPageAnnotations) {
+      webUrls.add(page.getUrl());
+    }
+
+    List<String> textInImage = new ArrayList<>();
+    for (EntityAnnotation textBlock : textAnnotations) {
+      textInImage.add(textBlock.getDescription());
+    }
+
+    List<ImageLabel> dominantColors = new ArrayList<>();
+    for (ColorInfo color : colorAnnotations) {
+      String nearestColor = getNearestColor(convertToJavaColor(color.getColor()));
+      dominantColors.add(new ImageLabel(nearestColor, color.getScore()));
+    }
+
+    List<ImageLabel> objectsInImage = new ArrayList<>();
+    for (LocalizedObjectAnnotation object : objectAnnotations) {
+      objectsInImage.add(new ImageLabel(object.getName(), object.getScore()));
+    }
+
+    List<ImageLabel> logosInImage = new ArrayList<>();
+    for (EntityAnnotation logo : logoAnnotations) {
+      logosInImage.add(new ImageLabel(logo.getDescription(), logo.getScore()));
+    }
+
+    String imageURL = getUploadedFileUrl(imagesService, blobKey);
+    return gson.toJson(new CloudVisionAnnotation(imageURL,
+                                                 cleanUpLabels(genericLabels),
+                                                 cleanUpLabels(webLabels),
+                                                 webBestLabels,
+                                                 webUrls,
+                                                 textInImage,
+                                                 cleanUpLabels(dominantColors),
+                                                 cleanUpLabels(objectsInImage), 
+                                                 cleanUpLabels(logosInImage)));
+  }
+
+  // This cleans up a imageLabel list by removing duplicates and blank labels.
+  // Also remove labels with a score of less than 40% (this can be adjusted later)
+  public static List<ImageLabel> cleanUpLabels(List<ImageLabel> labels) {
+    List<ImageLabel> cleanedList = new ArrayList<>();
```

Comment:
You could also consider using a set for this, and then it would do the deduplication for you.

Reply:
I would need to defined my own hashCode() and equals() methods for the imageLabel object before I can use hashsets made of imageLabel objects. That way, it would properly recognize two objects as "equal" (having the same description). I guess it would be cleaner to do it that way, but I had chosen this method as it is slightly simpler. Is there a good way to write a hash function for my ImageLabel class?

Reply:
It's not uncommon to define your own equals() and hashCode() type things (not only in Java, you'll probably come across it in C++ and other languages as well).

If you only care about the description and not the score, your hashcode can just be the hash of the description. That way you might get some undefined behavior if you have to compare two ImageLabels with the same description and different scores, although I think that wouldn't happen in this case?